js_modules/dagit/packages/core/src/runs/useQueryPersistedLogFilter.ts

```diff
@@ -74,10 +76,27 @@ export function encodeRunPageFilters(filter: LogFilter) {
   };
 }
 
-export function useQueryPersistedLogFilter() {
+export const EnabledRunLogLevelsKey = 'EnabledRunLogLevels';
+
+export const validateLogLevels = (json: any) => {
+  if (json === undefined || !Array.isArray(json)) {
+    return [];
+  }
+
+  const validLevels = new Set(Object.keys(LogLevel));
+  return json.filter((level) => validLevels.has(level));
+};
+
+export function useQueryPersistedLogFilter(): [LogFilter, (updates: LogFilter) => void] {
+  const [storedLogLevels] = useStateWithStorage(EnabledRunLogLevelsKey, validateLogLevels);
+
+  const defaults = React.useMemo(() => {
+    return {...DefaultQuerystring, levels: levelsToQuery(storedLogLevels || DefaultLogLevels)};
+  }, [storedLogLevels]);
+
```

Comment:
I think this hook is only consumed in the LogsToolbar so it probably doesn't matter, but I could see moving the setter that copies the selection to local storage into this hook (and exposing it in place of the setter returned by on line 97) so that the `useStateWithStorage(EnabledRunLogLevelsKey)` can be private to this file.

Reply:
This was my initial approach. The problem is that it can end up writing levels to storage based on the _querystring_, which I don't think we actually want.

That is, if I navigate to a run with levels set in the querystring, I want those levels to be applied to the filter because that's what I'm being permalinked to, but I *don't* want them written to storage -- they aren't my choices. I think we only want to write to storage when the user explicitly chooses filters.

So within this hook, we only want to read the persisted values and treat them as defaults, and we leave the setter behavior to the component usage itself. I'll add a comment to that effect.

Reply:
Ahh ok that makes sense!